File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sol(TreeNode* root){
        if(!root){return NULL;}
        
        TreeNode* new_r=root->right;
        if(root->left){
            TreeNode* nR=sol(root->left);
            root->right=root->left;
            root->left=NULL;
            if(nR)nR->right=new_r;
        }
        
        TreeNode* returningR=sol(root->right);
        if(!root->left&&!root->right){return root;}
        return returningR;
        
    }
    void flatten(TreeNode* root) {
        sol(root);
        
    }
};

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode sol(TreeNode root) {
        if (root == null) {
            return null;
        }

        TreeNode newRight = root.right;
        if (root.left != null) {
            TreeNode nR = sol(root.left);
            root.right = root.left;
            root.left = null;
            if (nR != null) {
                nR.right = newRight;
            }
        }

        TreeNode returningRight = sol(root.right);
        if (root.left == null && root.right == null) {
            return root;
        }
        return returningRight;
    }

    public void flatten(TreeNode root) {
        sol(root);
    }
}
